x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/allocation/TrainedModelAllocationClusterService.java

```diff
@@ -0,0 +1,376 @@
+import org.elasticsearch.xpack.core.ml.inference.allocation.TrainedModelAllocation;
+import org.elasticsearch.xpack.ml.MachineLearning;
+import org.elasticsearch.xpack.ml.job.NodeLoad;
+import org.elasticsearch.xpack.ml.job.NodeLoadDetector;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class TrainedModelAllocationClusterService implements ClusterStateListener {
+
+    private static final Logger logger = LogManager.getLogger(TrainedModelAllocationClusterService.class);
+
+    private final ClusterService clusterService;
+    private final NodeLoadDetector nodeLoadDetector;
+    private volatile int maxMemoryPercentage;
+    private volatile boolean useAuto;
+
+    public TrainedModelAllocationClusterService(Settings settings, ClusterService clusterService, NodeLoadDetector nodeLoadDetector) {
+        this.clusterService = clusterService;
+        this.nodeLoadDetector = nodeLoadDetector;
+        this.maxMemoryPercentage = MachineLearning.MAX_MACHINE_MEMORY_PERCENT.get(settings);
+        this.useAuto = MachineLearning.USE_AUTO_MACHINE_MEMORY_PERCENT.get(settings);
+        // Only nodes that can possibly be master nodes really need this service running
+        if (DiscoveryNode.isMasterNode(settings)) {
+            clusterService.addListener(this);
+            clusterService.getClusterSettings()
+                .addSettingsUpdateConsumer(MachineLearning.MAX_MACHINE_MEMORY_PERCENT, this::setMaxMemoryPercentage);
+            clusterService.getClusterSettings()
+                .addSettingsUpdateConsumer(MachineLearning.USE_AUTO_MACHINE_MEMORY_PERCENT, this::setUseAuto);
+        }
+    }
+
+    private void setMaxMemoryPercentage(int maxMemoryPercentage) {
+        this.maxMemoryPercentage = maxMemoryPercentage;
+    }
+
+    private void setUseAuto(boolean useAuto) {
+        this.useAuto = useAuto;
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
+        }
+        if (event.localNodeMaster() && shouldAllocateModels(event)) {
+            clusterService.submitStateUpdateTask("allocating models to nodes", new ClusterStateUpdateTask() {
+                @Override
+                public ClusterState execute(ClusterState currentState) {
+                    // TODO this has a weird side-effect for allocating to nodes
+                    // If the event indicates there were nodes added/removed, this method only looks at the current state and has
+                    // no previous knowledge of existing nodes. Consequently, if a model was manually removed (task-kill) from a node
+                    // it may get re-allocated to that node when another node is added/removed...
+                    return addRemoveAllocationNodes(currentState);
```

Comment:
I am not 100% sure the way around this. 

It may be that a node simply stays in the routing table with the `STOPPED` status to prevent this type of situation...